compiler/rustc_parse/src/lexer/tokentrees.rs

```diff
@@ -264,27 +262,19 @@ impl<'a> TokenTreesReader<'a> {
             }
             _ => {
                 let tt = TokenTree::Token(self.token.take());
-                self.real_token();
-                let is_joint = self.joint_to_prev == Joint && self.token.is_op();
-                Ok((tt, if is_joint { Joint } else { NonJoint }))
+                let mut is_joint = self.bump();
+                if !self.token.is_op() {
+                    is_joint = NonJoint;
+                }
```

Comment:
I originally forgot this condition, and that failed some proc-macro tests. This is a bit weird -- I would expect this jointness censoring to happen at the proc_macro_srv layer, and not here.

Reply:
Yes, this check should be moved to proc macro server.
Also lexer should produce the jointness flag for the "delimited group" token trees as well (the match branches above).

Reply:
Should I do this in this PR, or a separate one?

I've also rebased #75528 on top of this PR, seems to work

https://github.com/matklad/rust/commit/21bf6ce57289a50be8bced3abe44e23ca9ceeb56

Reply:
>Should I do this in this PR, or a separate one?

Whichever is more convenient.